lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php

```diff
@@ -114,31 +114,127 @@ public function __construct($dbh, $statement, OCI8Connection $conn)
+        $fragmentOffset = $tokenOffset = 0;
+        $fragments = $paramMap = array();
+        $currentLiteralDelimiter = null;
+
+        do {
+            if (!$currentLiteralDelimiter) {
+                $result = self::findPlaceholderOrOpeningQuote(
+                    $statement,
+                    $tokenOffset,
+                    $fragmentOffset,
+                    $fragments,
+                    $currentLiteralDelimiter,
+                    $paramMap
+                );
+            } else {
+                $result = self::findClosingQuote($statement, $tokenOffset, $currentLiteralDelimiter);
             }
+        } while ($result);
+
+        if ($currentLiteralDelimiter) {
+            throw new OCI8Exception(sprintf(
+                'The statement contains non-terminated string literal starting at offset %d',
+                $tokenOffset - 1
+            ));
         }
 
+        $fragments[] = substr($statement, $fragmentOffset);
+        $statement = implode('', $fragments);
+
         return array($statement, $paramMap);
     }
 
     /**
+     * Finds next placeholder or opening quote.
+     *
+     * @param string $statement The SQL statement to parse
+     * @param string $tokenOffset The offset to start searching from
+     * @param int $fragmentOffset The offset to build the next fragment from
+     * @param array $fragments Resulting query fragments
+     * @param string|null $currentLiteralDelimiter Current literal delimiter
+     * @param array $paramMap Parameter map
+     * @return bool Whether the token was found
+     */
+    private static function findPlaceholderOrOpeningQuote(
+        $statement,
+        &$tokenOffset,
+        &$fragmentOffset,
+        &$fragments,
+        &$currentLiteralDelimiter,
+        &$paramMap
+    ) {
+        $token = self::findToken($statement, $tokenOffset, '/[?\'"]/');
+
+        if (!$token) {
+            return false;
+        }
+
+        if ($token == '?') {
+            $position = count($paramMap) + 1;
+            $param = ':param' . $position;
+            $fragments[] = substr($statement, $fragmentOffset, $tokenOffset - $fragmentOffset);
+            $fragments[] = $param;
+            $paramMap[$position] = $param;
+            $fragmentOffset = ++$tokenOffset;
+        } else {
+            $currentLiteralDelimiter = $token;
+            ++$tokenOffset;
+        }
+
+        return true;
+    }
+
+    /**
+     * Finds closing quote
+     *
+     * @param string $statement The SQL statement to parse
+     * @param string $tokenOffset The offset to start searching from
+     * @param string|null $currentLiteralDelimiter Current literal delimiter
+     * @return bool Whether the token was found
+     */
+    private static function findClosingQuote(
+        $statement,
+        &$tokenOffset,
+        &$currentLiteralDelimiter
+    ) {
+        $token = self::findToken($statement, $tokenOffset, '/(?<!\\\\)' . $currentLiteralDelimiter . '/');
```

Comment:
It's a negative lookbehind: we're searching for the current literal delimiter which is not preceded by a backslash (i.e. not escaped). The backslash is escaped twice: once to avoid special meaning of the backslash a PHP string and once to avoid its special meaning in the regex.